enhancement: External validations for Operator SDK

jmrodri opened the pull request:
Today, validations used by CVP are compiled into the operator-sdk. Any changes 
to the validation rules requires a release of operator-framework/api followed by
a release of operator-sdk. This proposal attempts to design a way where the 
validations can be hosted in their own repos as well as updated without
requiring new releases of the operator-sdk.

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
```diff
@@ -0,0 +1,494 @@
+rules to require an immediate upstream operator-sdk release, which may
+otherwise be as far as 3 weeks away. It is difficult to explain to the community
+why we need a new release for a CVP need.
+
```
WDYT about add here as motivation:
- Able to have vendor-like validators and custom validators created by the users?

It seems for me the strongest motivation since today we cannot ship vendor-like ones in the operator-framework/api.

jmrodri replied:
@camilamacedo86 what do you mean by "vendor-like"?

camilamacedo86 replied:
Checks specifics for OCP k8s vendor for e.g.

jmrodri replied:
> Checks specifics for OCP k8s vendor for e.g.

OH! vendor *specific* validations. I will add that to the list. Thanks for clarifying.

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
IHMO we need to allow the output to be done in 2 formats at least.
1 - Using logger which is for the human reader. See that currently, it will log with colours the errors, the warns and Infos for users:

<img width="637" alt="Screenshot 2021-11-08 at 13 42 36" src="https://user-images.githubusercontent.com/7708031/140752417-378d5883-3ff6-46ea-9061-7e5c45e2c8e6.png">

2. - Then, the other option would be in JSON for machine consumptions. 

WDYT about for now we keep all as it is and just. move to operator-framework/api https://github.com/operator-framework/operator-sdk/blob/master/internal/cmd/operator-sdk/bundle/validate/internal/logger.go and https://github.com/camilamacedo86/ocp-olm-catalog-validator/blob/main/pkg/result/result.go to allow its re-usage?

For the future I'd like to propose:
- The structure that defines the output of the validator in the operator-framwork/api has the info level also. See:https://github.com/operator-framework/api/issues/158
- We centralize in the operator-framework/api the valid outputs formats and ensure that the validators and scorecard adopted the same standard to make it easier for consumers.

jmrodri replied:
> 1 - Using logger which is for the human reader. See that currently, it will log with colours the errors, the warns and Infos for users:

I didn't include that as I didn't intend for the validators to be run by themselves without `operator-sdk`. I do not disallow it but the sole purpose in my head was that they were to support being run by `operator-sdk`. I'll add this to the open questions to discuss more. I'm not opposed but it will change some assumptions.

camilamacedo86 replied:
@jmrodri, I am totally fine if the external one only outputs JSON format and SDK is able to consume that and make the result pretty as well if you see that is an easier path for us to start.

jmrodri replied:
> @jmrodri, I am totally fine if the external one only outputs JSON format and SDK is able to consume that and make the result pretty as well if you see that is an easier path for us to start.

Yes, that is definitely easier in the short term.

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
```suggestion
ManifestResult JSON to stdout.
```
should not it be ManifestResult?

jmrodri replied:
Yes, sorry about that.  I will update the EP.  https://github.com/operator-framework/api/blob/master/pkg/validation/errors/error.go#L9-L16

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
```suggestion
The validator should validate the given bundle and output the ManifestResult JSON.
```

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
```suggestion
Allowing the validators to output ManifestResult should make it easier to
```

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
```suggestion
output `ManifestResult `, it's easiest if we simply print that out as JSON to
```

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
The only problem is that we can use the validators as end-users who in POV would prefer to see the pretty results with logger.

jmrodri replied:
This is speaking directly about the validator executables which are to be run by SDK. They will output ManifestResult and the SDK will parse the information and output to the user a pretty logger format

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
Could you please clarify the XDG_CONFIG? How does it work?

jmrodri replied:
For now, the user will pass in the path to the validators using the `--alpha-select-external` flag. We will NOT auto discover them right now.

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
IHMO: https://github.com/operator-framework/enhancements/pull/98/files#r744739806
However, in the future, we might want to support other formats such as  JUnit or any other readable output.
Because of this in the medium term, maybe for SDK 2.0, I think that would be nice we adopted some standard define in the API for the validators and scorecard.

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
This one is vendor-like and we must remove it from operator-framework/API.
I'd like to propose we get started with an alpha version of the external validators to allow us:
- Remove these checks from the operator-framework/api
- Start to use and evaluate the solution for external validators

jmrodri replied:
I removed the list of existing validators. I didn't think they were adding any value to the EP.

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
IHMO the external validators ought to be responsible to test their rules. 
SDK just need to ensure and test that still able to work well with the external validators feature.

jmrodri replied:
Yes, removed the above generic template wording and replaced it with this:
```
* unit testing of the sdk external validator feature
* bundle validate e2e
* external validators will have their own test coverages
```

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
```suggestion
  * output from validator: `ManifestResults` JSON
```

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
WDYT about we remove the sections that we do not need to fill from the EP to try to make it less verbose?

camilamacedo86 commented:
I tried to contribute with some comments/nits.
However, otherwise, it seems amazing to me 🥇.

I think that can be very helpful for we are able to do specific checks and allow the users/community to create their own checks.

Terrific job @jmrodri!

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
We could use viper which allows flags and env vars: https://github.com/spf13/viper
That would be aligned with Kubebuilder plan: https://github.com/kubernetes-sigs/kubebuilder/issues/2388#issuecomment-945004547

jmrodri replied:
I decided to avoid the environment variable altogether. I went with an alpha level flag: `--alpha-select-external`

jmrodri commented:
Implementation https://github.com/operator-framework/operator-sdk/pull/5525

ryantking commented on enhancements/sdk-external-and-pluggable-validations.md:
```diff
@@ -0,0 +1,330 @@
+to the validation rules requires a release of operator-framework/api followed by
+a release of `operator-sdk`. This proposal attempts to design a way where the
+validations can be hosted in their own repos as well as updated without
+requiring new releases of the `operator-sdk`.
```
Can we add an example of a validator is currently defined in-code?

ryantking commented on enhancements/sdk-external-and-pluggable-validations.md:
Are we going to expose this as a go type that validator implementers will use?

jmrodri replied:
@ryantking yes, we are going to expose it. There is a link to it in the EP: https://github.com/operator-framework/api/blob/master/pkg/validation/errors/error.go#L9-L16

That's already accessible from SDK.

jmrodri replied:
Let me know if I need to call that out better.

ryantking commented on enhancements/sdk-external-and-pluggable-validations.md:
Will we support validator scripts?

```
python /home/user/dev/gatekeeper-operator/validator.py /home/user/dev/gatekeeper-operator/bundle
```

If not I think we should specifically call out that scripts complex CLI tools with flags and all that cannot be used as of this EP, only single executables that take the bundle path as a CLI argument.

jmrodri replied:
Great call out. The validators can be written in any language but it needs to have a single executable. So if we did it in python you'd have to do a `chmod +x on validator.py` so that it can be run by this feature.

jmrodri replied:
I added the following to this section:
```
The validators can be written in any language but there must be an executable
entry point that accepts a single bundle path as a CLI argument. For example,
you can write your validator in python but you would want to make the main 
python file executable or supply a shell script to be invoked.
```

ryantking commented on enhancements/sdk-external-and-pluggable-validations.md:
One risk I would call out is that users who run operators in an airgapped environment will have to add additional dependencies in the form of these validator executables. I assume the mitigation is they will compile them in the current way?

jmrodri replied:
They would not be able to compile them in themselves without forking operator-sdk and building it themselves.  I will add something about possibly adding a way to install these more easily from the operator-sdk binary itself for a future version.

ryantking reviewed (commented):
Generally looks pretty good, just a few questions and things I think would improve the reading of the EP from someone with less context.

I opted to leave out any formatting issues, but there's some funky numbering and capitalization stuff going on. I'd proofread quickly then blast the document with prettier.

jmrodri commented on enhancements/sdk-external-and-pluggable-validations.md:
```diff
@@ -0,0 +1,330 @@
+## Open Questions [optional]
+
+ 1. can we use scorecard to replace all of these validations?
+    * scorecard uses the cluster to run the tests, these validations typically
+      run locally or in a pipeline. They are also done before the expensive
+      operator tests are run.
```
@ryantking Most of the validations are pretty static. We could revisit this idea of enhancing scorecard for this purpose in the future.

camilamacedo86 replied:
@ryantking, 

The Scorecard tests require to be executed with a cluster and have a high cost to be executed.  Its initial idea would also allow functional tests and the users to create their own custom tests. 

See that for example, we check all bundles using SDK in the Audit tool (https://github.com/operator-framework/audit), by using the scorecard we need like 4 hours when doing exactly the same with the static validators it takes 1 hour.

jmrodri commented:
@ryantking I think I addressed your comments. Let me know if you want me to add a more explicit link to the `ManifestResult` struct or anything else.

jmrodri commented:
I will add the validator example on Monday.

jmrodri commented:
@ryantking I don't see the funky formatting when I view the rendered file in github:
[validatorn-ep.pdf](https://github.com/operator-framework/enhancements/files/8006297/validatorn-ep.pdf)

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
We have no 2 external validators that can be used to be tested checked with SDK: 
- https://github.com/redhat-openshift-ecosystem/ocp-olm-catalog-validator
- https://github.com/k8s-operatorhub/bundle-validator

jmrodri replied:
I will update the examples.

jmrodri replied:
@camilamacedo86 so did we make these to be used by the external validator feature? or they were stand-alone ones?

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
```suggestion
    * would need a cluster to run these validations
    * by requiring a cluster the tests require like 3 minutes to be executed instead of milliseconds which make it very hard we are able to re-test all bundle of an index as we do via Audit Tool for example. 
```

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
```suggestion
    * would have to write the engine to know how to execute these
    * the validators checks are static checks with a very low cost to be executed. By
```

jmrodri replied:
this is a repeat and I don't think it needs to be added as a con to the JavaScript or CUE alternative. These would be equally fast.

camilamacedo86 commented on enhancements/sdk-external-and-pluggable-validations.md:
Currently, we do not have many external validators. However, in the future, if it grows we might face complexities such as:

To test your project against the criteria to publish in the index X:

- download external validator A, B, C
- test your bundle against `operator-sdk bundle validate ./bundle --select-optional suite=operatorframework` and then against A,B,C validators as well. 

So, shows that would be very nice if we could just copy and paste a config and run just 1 command that would check the bundle against the common criteria + all external validators. 

See that we download OPM CLI and use it (https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v3/memcached-operator/Makefile#L183-L198). Then, if we could make SDK download and use all external validators configured then, we maybe make it easier / possible SDK only uses this config with a bin provided by operator-framework/api instead of importing it to execute the checks. That could address the main motivations of this PE easily. 

Then, if a user requires a new version of the default checks (of/api) we could only update the config for its upper version as we do in the case of OPM in the makefile target.

jmrodri replied:
This config is going down the route of "discoverability".

jmrodri commented on enhancements/sdk-external-and-pluggable-validations.md:
```diff
@@ -0,0 +1,376 @@
+
+#### Story 5
+* Users can define the target set of validation rules for SDK's bundle validation command (either from local or remote source).
+
```
@camilamacedo86 I still think there is value for the compiled in checks. Though we may have to think of a better way to update those.

jmrodri commented on enhancements/sdk-external-and-pluggable-validations.md:
@camilamacedo86 `pkg/result` shouldn't be in the external validators. I do believe we can have a future enhancement to consolidate some of the logic particular logic on opening and working through bundles.

jmrodri commented:
@camilamacedo86 if you want discoverability to be part of this EP, let me know so we can discuss how we think we should do that. @ryantking is currently working on some validators.

jmrodri commented:
/hold cancel

jmrodri commented:
@camilamacedo86 @ryantking I'm merging this EP. If we need to add more we can definitely enhance this enhancement (just updating the same one would be fine).
